File: web-app/backend/app/routers/exposure.py
from collections import defaultdict

from fastapi import APIRouter

from app.schemas.exposure import ExposureItem, ExposureResponse
from app.sheets.performance import fetch_performance

router = APIRouter()

# 対象通貨（HKD は除外）
TARGET_CURRENCIES = {"JPY", "USD"}


@router.get("/exposure", response_model=ExposureResponse)
async def get_exposure() -> ExposureResponse:
    """通貨別エクスポージャーサマリーを返す（最新月、JPY/USD のみ）"""
    records = fetch_performance()
    if not records:
        return ExposureResponse(items=[])

    # 最新月を特定
    latest_date = max(r["date"] for r in records)
    latest_records = [r for r in records if r["date"] == latest_date]

    # 通貨別に集計
    agg: dict[str, dict] = defaultdict(lambda: {"value": 0.0, "cost": 0.0})
    for r in latest_records:
        currency = str(r.get("currency", "JPY"))
        if currency not in TARGET_CURRENCIES:
            continue
        agg[currency]["value"] += r["value"]
        agg[currency]["cost"] += r["cost"]

    total_value = sum(v["value"] for v in agg.values())

    items = []
    for currency in sorted(agg.keys()):
        v = agg[currency]
        profit = v["value"] - v["cost"]
        profit_rate = profit / v["cost"] * 100 if v["cost"] else 0.0
        percentage = v["value"] / total_value * 100 if total_value else 0.0
        items.append(
            ExposureItem(
                currency=currency,
                value=round(v["value"], 0),
                cost=round(v["cost"], 0),
                profit=round(profit, 0),
                profitRate=round(profit_rate, 2),
                percentage=round(percentage, 2),
            )
        )
    return ExposureResponse(items=items)

File: portfolio-dashboard/server/src/routes/exposure.ts
import { Hono } from "hono";
import { getLatestDate, getLatestPnlRecords } from "../db/queries.js";

// 対象通貨（HKD 除外）
const TARGET_CURRENCIES = ["JPY", "USD"];

const app = new Hono();

app.get("/", (c) => {
  const latestDate = getLatestDate();

  if (!latestDate) {
    return c.json({ items: [] });
  }

  const records = getLatestPnlRecords(latestDate);

  // JPY / USD のみに絞り込んで通貨別に集計
  const currencyMap = new Map<string, { value: number; cost: number }>();

  for (const r of records) {
    if (!TARGET_CURRENCIES.includes(r.currency)) continue;
    const acc = currencyMap.get(r.currency) ?? { value: 0, cost: 0 };
    acc.value += r.value;
    acc.cost += r.cost;
    currencyMap.set(r.currency, acc);
  }

  const totalValue = [...currencyMap.values()].reduce(
    (sum, v) => sum + v.value,
    0,
  );

  const items = [...currencyMap.entries()]
    .sort(([a], [b]) => a.localeCompare(b)) // アルファベット順
    .map(([currency, { value, cost }]) => {
      const profit = value - cost;
      const profitRate = cost > 0 ? (profit / cost) * 100 : 0;
      const percentage = totalValue > 0 ? (value / totalValue) * 100 : 0;

      return {
        currency,
        value: Math.round(value),
        cost: Math.round(cost),
        profit: Math.round(profit),
        profitRate: Math.round(profitRate * 100) / 100,
        percentage: Math.round(percentage * 100) / 100,
      };
    });

  return c.json({ items });
});

export { app as exposureRoute };
